src/crimson/osd/replicated_recovery_backend.cc

```diff
@@ -306,7 +306,10 @@ ReplicatedRecoveryBackend::recover_delete(
       }
       return seastar::make_ready_future<>();
     }).then_interruptible([this, soid, &stat_diff] {
-      pg.get_recovery_handler()->on_global_recover(soid, stat_diff, true);
+      const auto &missing = pg.get_peering_state().get_pg_log().get_missing();
+      if (!missing.is_missing(soid)) {
+        pg.get_recovery_handler()->on_global_recover(soid, stat_diff, true);
```

Comment:
We already have a `if (!object_missing) {` case right before, can we call on global recover there?
```
	if (!object_missing) {
	  stat_diff.num_objects_recovered = 1;
+         pg.get_recovery_handler()->on_global_recover(soid, stat_diff, true);
	  return seastar::make_ready_future<>();
	} 
```

Reply:
I think `on_global_recover()` needs to be called after `push_delete()`; and the earlier `object_missing` is an indication of whether the object is missing on any replica osds, and the check here is whether it's missing on the primary.